ledger-service/http-json/src/test/scala/com/digitalasset/http/util/ContractStreamStepTest.scala

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.http
+package util
+
+import com.digitalasset.daml.lf.data.FlatSpecCheckLaws
+
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.{FlatSpec, Matchers}
+import scalaz.scalacheck.ScalaCheckBinding._
+import scalaz.scalacheck.ScalazProperties
+import scalaz.syntax.apply._
+import scalaz.syntax.semigroup._
+import scalaz.{@@, Equal, Semigroup, Tag}
+
+class ContractStreamStepTest
+    extends FlatSpec
+    with FlatSpecCheckLaws
+    with Matchers
+    with GeneratorDrivenPropertyChecks
+    with TableDrivenPropertyChecks {
+  import ContractStreamStepTest._, ContractStreamStep._
+  import InsertDeleteStepTest._
+
+  behavior of "append"
+
+  it should "be associative for valid streams" in forAll(validStreamGen) { csses =>
+    whenever(csses.size >= 3) {
+      forEvery(
+        Table(("a", "b", "c"), csses.sliding(3).map { case Seq(a, b, c) => (a, b, c) }.toSeq: _*)) {
+        case (a, b, c) =>
+          (a |+| (b |+| c)) should ===((a |+| b) |+| c)
+      }
+    }
+  }
+
+  it should "report the last offset" in forAll { (a: CSS, b: CSS) =>
+    def off(css: ContractStreamStep[_, _]) = css match {
+      case Acs(_) => None
+      case LiveBegin(off) => off.toOption
+      case Txn(_, off) => Some(off)
+    }
+    off(a |+| b) should ===(off(b) orElse off(a))
+  }
+
+  it should "preserve append across toInsertDelete" in forAll { (a: CSS, b: CSS) =>
+    (a |+| b).toInsertDelete should ===(a.toInsertDelete |+| b.toInsertDelete)
+  }
+
+  behavior of "append semigroup"
+
+  checkLaws(ScalazProperties.semigroup.laws[CSS])
+}
+
+object ContractStreamStepTest {
+  import InsertDeleteStepTest._, InsertDeleteStep.Inserts, ContractStreamStep._
+  import org.scalacheck.{Arbitrary, Gen}
+  import Arbitrary.arbitrary
+
+  type CSS = ContractStreamStep[Unit, Cid]
+
+  implicit val `CSS semigroup`: Semigroup[CSS] =
```

Comment:
Should it be part of `ContractStreamStep` companion object, it is useful property.